File: dbassistant/tokenhandlers/GlossaryTokenHandler.py
from dbassistant.analysis.Logging import Logging
from dbassistant.exceptions.ArgumentHandlerIsFullException import ArgumentHandlerIsFullException
from dbassistant.exceptions.MishandledTokenException import MishandledTokenException
from dbassistant.interfaces.ITokenHandler import ITokenHandler
from dbassistant.tokens.Token import Token
from dbassistant.tokens.TokenType import TokenType

class GlossaryTokenHandler(ITokenHandler):
    def __init__(self, priority, classList, keyMap):
        self._priority = priority
        self.classList = classList # unique set of words
        self.keyMap = keyMap #with spaces

        # keymap with spaces replaced by underscores
        self.tokenMap = {}
        for key in keyMap.copy().keys():
            self.tokenMap[key.replace(" ", "_")] = self.keyMap[key]
    
    @property
    def priority(self):
        return self._priority
    
    @priority.setter
    def priority(self, value):
        self._priority = value

    @property
    def size(self):
        return len(self.classList) + len(self.keyMap.keys())
    
    @property
    def utilizedSize(self):
        return self.size
    
    @property
    def keyCount(self):
        return len(self.keyMap.keys())
    
    @property
    def mutable(self):
        return False
    

    #Encode with token map, because tokens must not contain spaces
    def canEncode(self, word):
        return word in self.classList or word in list(self.tokenMap)
    
    def encode(self, word):
        if word in self.classList:
            classListIndex = 0
            for _class in self.classList:
                if _class == word:
                    break
                classListIndex += 1
            return Token(classListIndex, TokenType.CLASS)
        elif word in list(self.tokenMap):
            cid = self.tokenMap[word]
            return Token(cid, TokenType.KEY)
        else:
            raise MishandledTokenException()
    
    #Decode with key map, because output should be humanly readable
    def decode(self, token):
        if token < len(self.classList):
            return list(self.classList)[token]
        elif token < len(self.classList) + len(self.keyMap.keys()):
            return list(self.keyMap.keys())[token - len(self.classList)]
        else:
            raise MishandledTokenException()
    
    def resetState(self):
        pass

    def serialize(self, path):
        file = path + "/glossary_tokens.layer"
        with open(file, "w") as f:
            f.write("GlossaryTokenHandler\n")
            f.write("{}\n".format(len(self.classList)))
            for _class in self.classList:
                f.write("{}\n".format(_class))
            f.write("{}\n".format(len(self.keyMap.keys())))
            for key in self.keyMap.keys():
                f.write("{}\n".format(key))
                f.write("{}\n".format(self.keyMap[key]))

    def deserialize(self,path):
        file = path + "/glossary_tokens.layer"
        with open(file, "r") as f:
            line = f.readline().strip()
            if line != "GlossaryTokenHandler":
                raise Exception("Wrong token handler type")
            classListSize = int(f.readline().strip())
            for _ in range(classListSize):
                _class = f.readline().strip()
                self.classList.add(_class)
            keyMapSize = int(f.readline().strip())
            for _ in range(keyMapSize):
                key = f.readline().strip()
                cid = int(f.readline().strip())
                self.keyMap[key] = cid
                self.tokenMap[key.replace(" ", "_")] = cid
        

File: dbassistant/tokenhandlers/GlossaryTokenHandler.ts
import Token from '../tokens/Token'; // Import Token and TokenType classes
import TokenType from '../tokens/TokenType'; // Import Token and TokenType classes
import ITokenHandler from '../interfaces/ITokenHandler'; // Import ITokenHandler interface
import TokenDecodeException from '../exceptions/TokenDecodeException';
import TokenEncodeException from '../exceptions/TokenEncodeException';

export class GlossaryTokenHandler implements ITokenHandler {
    private _priority: number;
    private classList: Set<string>;
    private keyMap: Record<string, number>;
    private tokenMap: Record<string, number>;

    constructor(priority: number, classList: Set<string>, keyMap: Record<string, number>) {
        this._priority = priority;
        this.classList = classList;
        this.keyMap = keyMap;

        this.tokenMap = {};
        for (const key of Object.keys(keyMap)) {
            this.tokenMap[key.replace(" ", "_")] = this.keyMap[key];
        }
    }

    get priority(): number {
        return this._priority;
    }

    set priority(value: number) {
        this._priority = value;
    }

    get size(): number {
        return this.classList.size + Object.keys(this.keyMap).length;
    }

    get utilizedSize(): number {
        return this.size;
    }

    get keyCount(): number {
        return Object.keys(this.keyMap).length;
    }

    get mutable(): boolean {
        return false;
    }

    canEncode(word: string): boolean {
        return this.classList.has(word) || Object.keys(this.tokenMap).includes(word);
    }

    encode(word: string): Token {
        if (this.classList.has(word)) {
            let classListIndex = 0;
            for (const _class of this.classList) {
                if (_class === word) {
                    break;
                }
                classListIndex++;
            }
            return new Token(classListIndex, TokenType.CLASS);
        } else if (Object.keys(this.tokenMap).includes(word)) {
            const cid = this.tokenMap[word];
            return new Token(cid, TokenType.KEY);
        } else {
            throw new TokenEncodeException(this, word);
        }
    }

    decode(token: Token): string {
        if (token.value < this.classList.size) {
            return Array.from(this.classList)[token.value];
        } else if (token.value < this.classList.size + Object.keys(this.keyMap).length) {
            const keys = Object.keys(this.keyMap);
            return keys[token.value - this.classList.size];
        } else {
            throw new TokenDecodeException();
        }
    }

    resetState(): void {
        // Implement resetState if needed
    }

    serialize(path: string): void {
        const file = `${path}/glossary_tokens.layer`;
        const content = `GlossaryTokenHandler\n${this.classList.size}\n${[...this.classList].join('\n')}\n${Object.keys(this.keyMap).length}\n`;
        
        for (const key of Object.keys(this.keyMap)) {
            content += `${key}\n${this.keyMap[key]}\n`;
        }

        // Write content to the file
    }

    deserialize(path: string): void {
        const file = `${path}/glossary_tokens.layer`;
        // Read the file and deserialize the content
    }
}
